torch/distributed/_composable/fsdp/_fsdp_param.py

```diff
@@ -0,0 +1,598 @@
+        # behavior (e.g. requiring casting); otherwise, clamp to `None`
+        if param_dtype is not None and param_dtype == self.orig_dtype:
+            # E.g. orig=compute=bf16
+            param_dtype = None
+        # By default, gradients are computed in the compute dtype
+        if reduce_dtype is not None and (
+            # E.g. orig=compute=reduce=bf16
+            (param_dtype is None and reduce_dtype == self.orig_dtype)
+            # E.g. orig=fp32, compute=reduce=bf16
+            or (param_dtype is not None and reduce_dtype == param_dtype)
+        ):
+            reduce_dtype = None
+        self.param_dtype = param_dtype
+        self.reduce_dtype = reduce_dtype
+        # None indicates that the mixed precision is not enabled
+
+    def _init_dtensor_attrs(self, param: nn.Parameter):
+        self._is_dtensor = isinstance(param, DTensor)  # assume TP
+        # These TP attributes are used for the unsharded parameter
+        if self._is_dtensor:
+            self._tp_spec = cast(DTensor, param)._spec
+            self._tp_global_size = param.size()
+            self._tp_global_stride = param.stride()
+
+    def _init_float8tensor_attrs(self, param: nn.Parameter):
+        self._is_float8tensor = getattr(param, "_is_fp8_weight", False)
+        module = self._module_info.module
+        # TODO: Leave this warning for now since we only integrate with
+        # float8-experimental. We need to figure out meta-device story, where
+        # tensor attributes are lost upon materializing on a real device.
+        if (
+            "Float8Linear" in str(type(module))
+            and "weight" in self._module_info.param_name
+        ) and not self._is_float8tensor:
+            warnings.warn("Float8Linear.weight has _is_float8tensor == False")
+        if self._is_float8tensor:
+            float8_scale_name = "fp8_scale_w"
+            float8_tensor_ctor_name = "_float8_tensor_ctor"
+            cast_w_to_float8_fn_name = "cast_w_to_float8"
+            is_amax_initialized_name = "is_amax_initialized"
+            float8_emulate_name = "emulate"
+            for attr_name in (
+                float8_scale_name,
+                float8_tensor_ctor_name,
+                cast_w_to_float8_fn_name,
+                is_amax_initialized_name,
+                float8_emulate_name,
+            ):
+                if not hasattr(module, attr_name):
+                    raise AssertionError(f"{type(module)} is missing {attr_name}")
+            self._float8_scale = getattr(module, float8_scale_name)
+            self._float8_tensor_ctor = getattr(module, float8_tensor_ctor_name)
+            self._cast_w_to_float8_fn = getattr(module, cast_w_to_float8_fn_name)
+            self._is_amax_initialized_attr_name = is_amax_initialized_name
+            self._float8_emulate_attr_name = float8_emulate_name
+
+    def _init_global_mesh_attrs(self, param: nn.Parameter):
```

Comment:
This function initializes:
```
_global_mesh
_global_placements
_global_size
_global_stride
```
which are used thereafter to wrap FSDP local tensors into `DTensor` (i.e. sharded parameter at construction time and sharded gradient each backward).

We can imagine this can be folded into `DTensor` / `DeviceMesh`:
(1) 1D FSDP: Given a `torch.Tensor`, construct a `DTensor` with `(Shard(0),)` placement with padded local tensor.
(2) 2D FSDP: Given a `DTensor`, construct a `DTensor` with _additional_ `(Shard(0),)` placement with padded local tensor.

Should we expose something similar to `empty_like` like `dtensor_like` for wrapping the sharded gradient? This is not necessary. The alternative is to read the attributes off the `sharded_param`, which is a DTensor with the correct metadata.